### ISSUE TITLE
layeredimage not displayed as intended when the same attribute is repeated

### ISSUE THREAD

**Issue by magicnutsgames** [renpy/renpy#3011]
Hi. I recently updated my Ren'py from 7.3.5 to 7.4.8 and noticed that when showing layered images, repeating the same attribute twice messes up the images, something that didn't cause any issues in the previous version.
For example:

`show lana mouthsmile mouthsmile eyesclosed`

The repetition of the smiling mouth in this case will cause the layered image to display the 'default' attribute for all of the groups that have a default attribute set for them (in addition to showing the repeated attribute). Going with this example, the character Lana will have her smiling mouth shown correctly, but the closed eyes will not be shown and instead, the sprite's neutral eyes will be displayed. 

[image files](https://whatalegendgame.com/wp-content/uploads/2021/09/pov.zip)
The layeredimage:

```
layeredimage pov:
    always:
        "pov/povnull.webp"

    group extraback:
        attribute xbshadowhwall2
        attribute xbshadowhwall1
        attribute xbnone:
            "null"

    group pants:
        attribute pneu default:
            "pov_base"
        attribute pdown0
        attribute pdown1
        attribute pdown2_1
        attribute pdown2_2
        attribute pdown3
        attribute ptopless

    group erect:
        attribute erect0:
            "null"
        attribute erect1
        attribute erect2
        attribute erect1topless
        attribute erect2topless

    group eyes auto:
        attribute eneu default
        attribute eclosed
        attribute ewrite
        attribute ecross
        attribute ecrossdown
        attribute ecrossup
        attribute eflat
        attribute eflatb
        attribute eflatc
        attribute ehappy
        attribute enarrow
        attribute enarrowb
        attribute enarrowc
        attribute eneub
        attribute eneubdown
        attribute eneubup
        attribute eneuc
        attribute eneucdown1
        attribute eneucdown2
        attribute eneucup
        attribute eneudown
        attribute eneuup
        attribute eshut
        attribute ewide
        attribute ewideb
        attribute ewidebdown
        attribute ewidebup
        attribute ewidec
        attribute ewidecdown
        attribute ewidecup
        attribute ewidedown
        attribute ewideup
        attribute ewink
        attribute ewinkc
        attribute ewinkdown
        attribute ewinkbdown
        attribute ethink
        attribute ethinknarrow
        attribute ethinkflat
        attribute ethinkwide
        attribute ethinkdown
        attribute ethinkwidedown
        attribute esides
        attribute esideswide
        attribute esidesnarrow
        attribute esidesdown
        attribute esideswidedown
        attribute ebag
        attribute eyawn
        attribute echarwide
        attribute echarwidec
        attribute echarwidedown
        attribute echarclosed
        attribute echeck
        attribute echeckwide
        attribute ecrystal
        attribute ecrystalfall
        attribute echeckobject
        attribute esidescheckobject
        attribute eneu_noblink:
            "pov/pov_eyes_eneu.webp"
        attribute ewide_noblink:
            "pov/pov_eyes_ewide.webp"
        attribute enarrow_noblink:
            "pov/pov_eyes_enarrow.webp"
        attribute enone:
            "null"
        attribute echeckup


    group brows:
        attribute bneu default
        attribute bangry
        attribute bdoubt
        attribute bsad
        attribute bup
        attribute bsus
        attribute bshrug
        attribute bthink
        attribute bcharm
        attribute bnone:
            "null"

    group mouth:
        attribute mneu default
        attribute mcharm
        attribute mgrin
        attribute mcry
        attribute mjawdrop
        attribute mhappy
        attribute mno
        attribute mpuff
        attribute msad
        attribute mside
        attribute msmile
        attribute mtight
        attribute mtongue
        attribute mwhistle
        attribute meat
        attribute mshout
        attribute mwow:
            "pov/pov_mouth_mwhistle1.webp"
        attribute mwonder
        attribute mopen:
            "pov/pov_mouth_mtalk2.webp"
        attribute mdislike
        attribute mvomit
        attribute mchew
        attribute myawn
        attribute mkiss
        attribute mkiss1
        attribute mkiss2
        attribute msilent:
            "pov/pov_mouth_mneu.webp"
        attribute msilenthappy:
            "pov/pov_mouth_mhappy.webp"
        attribute msilentsmile:
            "pov/pov_mouth_msmile.webp"
        attribute msilentgrin:
            "pov/pov_mouth_mcharming.webp"
        attribute msilentdislike:
            "pov/pov_mouth_mdislike.webp"
        attribute msilentsad:
            "pov/pov_mouth_msad.webp"
        attribute msilentcry:
            "pov/pov_mouth_mcry.webp"
        attribute msilentwonder:
            "pov/pov_mouth_mwonder.webp"
        attribute msilentside:
            "pov/pov_mouth_mside.webp"
        attribute mnone:
            "null"

    group skin:
        attribute sneu default:
            "null"
        attribute sblue
        attribute spurple
        attribute sblush
        attribute sbooger
        attribute scry
        attribute stear
        attribute sdirty
        attribute sgreen
        attribute sscratch1
        attribute sscratch2
        attribute sscratch3
        attribute sslap1
        attribute sslap2
        attribute ssweat
        attribute smud
        attribute sblushsweat
        attribute scrazyhair
        attribute scrazyhairsweat
        attribute scrazyhairblush
        attribute scrazyhairblushsweat
        attribute scrazyhairbump
        attribute scrazyhairbumppale
        attribute scrazyhairscratch
        attribute scrazyhairdust
        attribute spale
        attribute shungry
        attribute sdust


    group hands:
        attribute hnone default:
            "null"
        attribute hshrug:
            pos(38,222)
        attribute hbag
        attribute hobjectboxjack
        attribute hobjectmap
        attribute hobjectmapdrop
        attribute hpdownbag
        attribute hwrite
        attribute hwrite1
        attribute hcoverpenis
        attribute hwall
        attribute hwall1
        attribute hwall2
        attribute hfox
        attribute hboobmyrtle:
            pos(232,0)
        attribute hshirtoff:
            pos(-30, -11)

    group hug:
        attribute hugnone:
            "null"
        attribute hugmyrtle
        attribute hugmyrtlemessy
        attribute hugmyrtlepre
        attribute hugmyrtlepremessy
        attribute hugsimmonepre
        attribute hugsimmone
        attribute hugsimmonemessy
        attribute huglana

    group kiss:
        attribute kissnone:
            "null"
        attribute kissmyrtle
        attribute kissmyrtlemessy
        attribute kissmyrtlepre
        attribute kissmyrtlepremessy
        attribute kisssimmone
        attribute kisssimmonemessy
        attribute kisslana


    group handf:
        attribute hfneu1 default
        attribute hfneu2
        attribute hfhide
        attribute hfhead
        attribute hfreach
        attribute hfreach1
        attribute hfreach2
        attribute hfstop
        attribute hffist
        attribute hfpoint
        attribute hfthumb
        attribute hfsmall
        attribute hfnneu1
        attribute hfnneu2
        attribute hf0neu1
        attribute hf0neu2
        attribute hfcharneu1
        attribute hfcharneu2
        attribute hfcharreach
        attribute hfchartouch
        attribute hfshoulder
        attribute hftorch
        attribute hfnfloat:
            pos(-30, -11)
        attribute hfobjectring
        attribute hfobjectshear
        attribute hfobjectsalt
        attribute hfobjectroastrabbit
        attribute hfobjectrabbit
        attribute hfobjectpaper
        attribute hfobjectlog
        attribute hfobjectharness
        attribute hfobjectflower
        attribute hfobjectflowercrown
        attribute hfobjectdress
        attribute hfobjectwhitewool
        attribute hfobjectberry
        attribute hfobjectgluejar
        attribute hfobjectglue
        attribute hfobjectsapjar
        attribute hfobjectcookie
        attribute hfobjectsap
        attribute hfobjectmirror
        attribute hfobjectmirrorcharr
        attribute hfobjectmirrorface
        attribute hfobjectcrystal
        attribute hfobjectcrystalbroken
        attribute hfobjectcrystalfallup
        attribute hfobjectcrystalfall
        attribute hfobjectshard
        attribute hfobjectwetcake
        attribute hfobjectwetcakebite
        attribute hfobjectfish
        attribute hfobjectfishtoss
        attribute hfobjectfishtosspose
        attribute hfobjectglasses
        attribute hfobjectgold
        attribute hfobjectbag:
            pos(-311,0)
        attribute hfobjectad
        attribute hfobjectrope
        attribute hfobjectmop
        attribute hfobjectdildosmall
        attribute hfobjectmap
        attribute hfobjectmapfolded
        attribute hfobjectdildo:
            pos(-3,0)
        attribute hfobjectdildodown
        attribute hfobjectpillow:
            pos(-118,272)
        attribute hfobjectdecoy:
            pos(-118,272)
        attribute hfglasses
        attribute hfglassespose
        attribute hfglassesneu
        attribute hfshock
        attribute hfwhitefinger
        attribute hfsuckfinger
        attribute hfobjecttea
        attribute hfobjectbox
        attribute hfobjectbox1
        attribute hfobjectaxe
        attribute hfobjectbullskull:
            pos(-38, 200)
        attribute hfobjectsheepdress:
            pos(-38, 200)
        attribute hfobjectsheepdressmacho:
            pos(-38, 200)
        attribute hfobjectshirt:
            pos(-30, -11)
        attribute hfobjectshirtwet:
            pos(-30, -11)
        attribute hfstill1:
            "pov/pov_handf_hfneu1.webp"
        attribute hfstill2:
            "pov/pov_handf_hfneu2.webp"
        attribute hfpose:
            "pov/pov_handf_hftalk2.webp"
        attribute hfquestion:
            pos(38,222)
        attribute hfpixie:
            "pov/pov_handf_hftalk2.webp"
        attribute hfnone:
            "null"
        attribute hfnstill1:
            "pov/pov_handf_hfnneu1.webp"
        attribute hfhugmyrtle
        attribute hfhugmyrtlewaist
        attribute hfhugmyrtleass
        attribute hfkissmyrtle
        attribute hfkissmyrtlewaist
        attribute hfkissmyrtleass
        attribute hfobjectfeather

    group erectn:
        attribute erectn1
        attribute erectn2
        attribute erectn3

    group handb:
        attribute hbneu1 default
        attribute hbneu2
        attribute hbchest
        attribute hbface
        attribute hbbag
        attribute hbpdownbag
        attribute hbfront
        attribute hbthink
        attribute hbyawn
        attribute hbhide
        attribute hbnneu1
        attribute hbnneu2
        attribute hb0neu2
        attribute hbnfloat
        attribute hbraise
        attribute hbwrite
        attribute hbcharneu1
        attribute hbcharneu2
        attribute hbobjectgluepotion
        attribute hbobjectgluepotiondrink
        attribute hbobjectmushroom
        attribute hbobjectmushroomeaten
        attribute hbobjectletter
        attribute hbobjectdildoanal
        attribute hbobjectletteropen
        attribute hbquestion:
            pos(38,222)
        attribute hbnone:
            "null"
        attribute hbshock:
            pos(550,222)
        attribute hbhugmyrtle
        attribute hbkissmyrtle
        attribute hbboobmyrtle
        attribute hbobjectmap
```

**Comment by Gouvernathor** [comment]
This behavior has never been documented and therefore was never guaranteed to be kept in renpy.
Maybe you can do the same thing by doing `show lana as lana` in between, which may just reset lateredimage attributes while keeping the applied transforms (I didn't test it yet).

If it doesnt work, an argument could be made for a new keyword to be added to layeredimages (say, `reset`) which would revert all groups to their default attributes.

**Comment by magicnutsgames** [comment]
The fact that all groups revert or reset back to their default one (if they have any) once there's a repetition of an attribute is the problem.  

In the script below, for example, because I have written smiling mouth twice (msmile) -- something that happens by accident quite often when showing thousands of displayables, the eyes (ewide), skin (sblush), eyebrows (bup), and hands (hfneu2 and hbneu2), are all reset to the default attribute of their groups and except the mouth attribute (which was the one repeated twice), none of what I wanted to be shown is actually shown:

`show pov msmile bup ewide sblush hfneu2 hbneu2 msmile
`
Worst still, it doesn't even give an error to make you know something has gone wrong. So a player seeing the image might just think we messed it up by showing a bad image that doesn't make sense, rather than it being an unintended displayable. It also makes debugging much more difficult.

If you show relatively few layeredimages, the current system may work well since there is less chance of human error, but once you have thousands of instances of showing iayeredimages, or layerdimages with many groups, that once again increase the chance of human error, then the current issue is a major one, in my opinion. 

In my case, it has entirely messed up my game since it wasn't an issue in the previous version, and I have to check tens of thousands of displayables one by one to make sure they are correct. I sincerely doubt I'm alone in this, even if this problem has not been reported before.

**Comment by Gouvernathor** [comment]
Ah indeed, I misunderstood.
Fundamentally you're not supposed to double attributes like this, so I don't believe "as intended" in the title of the issue to be the most accurate term.

More on point, I tried defining the layeredimage with the code and assets you provided, and with the simple line `show pov msmile bup ewide sblush hfneu2 hbneu2 msmile` the character changes its blush, arms position, eyes, brows...
So, in a few words, can't replicate (in the last nightly).

**Comment by mal** [comment]
This appears to be a direct result of the change to treat attributes as ordered lists instead of unordered sets, introduced in `7.4.0`: https://github.com/renpy/renpy/commit/fcf85a59b7206d0bb9c70c55256571c30a7e9907. Reverting for the purposes of validation allows the test case below to work without issue in 7.4.8, but is not a solution due to dependency on this change by newer features such as - I believe - Live 2D support and possibly others.

Minimal testcase which works fine on 7.3.5, and surfaces the issue on 7.4.8.

```rpy
label main_menu:
    $ _confirm_quit = False
    return


layeredimage li:
    always:
        Fixed(Solid('0004', align=(.5, .5), xysize=(800, 400)))

    group lhs:
        align (.25, .5)
        attribute lr default Solid('f778', xysize=(320, 240))
        attribute lb Solid('7ff8', xysize=(320, 240))
        attribute ly Solid('ff78', xysize=(320, 240))

    group rhs:
        align (.75, .5)
        attribute rr default Solid('f778', xysize=(320, 240))
        attribute rb Solid('7ff8', xysize=(320, 240))
        attribute ry Solid('ff78', xysize=(320, 240))

    group test:
        align (.5, .3)
        attribute flag default Text('Default applied to test group!')
        attribute clear Null(size=10)


label start:
    scene
    show li clear at truecenter
    'red - red'
    show li lb ry
    'blue - yellow'
    show li ly rb
    'yellow - blue'
    show li lb ry ry
    'blue - yellowyellow'
    return
```

**Comment by Gouvernathor** [comment]
Maybe an OrderedSet chould be used. The type doesn't exist per se, but an OrderedDict whose keys would be set to None could be used, then converted to a list if necessary to avoid confusion.
Or any other way of removing doubles which would not depend on the creation of sets.

That is, if the only goal was to keep the initial order. But there could be an obscure use case where stating the same attribute twice can mean something. I doubt it since the use of sets seemed to be part of the 7.3 release code, but well...

**Comment by renpytom** [comment]
A doubled attribute should not mean anything to a layered image. I

We'll fix the layered image implementation to use a set.

**Comment by Gouvernathor** [comment]
I added this to the beginning of layeredimage's `_choose_attributes` method :
```py
attributes = list(set(attributes))
optional = list(set(optional))
```
I expected it to solve this issue (at the cost of maybe creating others), but it didn't.
Expectedly, turning the lists into sets in the last part of fcf85a5 had the same non-result.
The issue comes from the first part, from the `apply_attributes` method of the ShowImageInfo class.
Therefore, I don't see the solution coming from a change in layeredimage implementation.

**Comment by renpytom** [comment]
Fixed by #3015.
